### PR TITLE
Added dynamic text positioning 

### PR DESCRIPTION
**Pull request by PaulNicolasHunter**
This is an improvement to my PR #188.

I've added dynamic text positioning to the LineSlider2D continuing the reference of PR #178.
There is a new parameter "alignment" which puts the test position to the right/left (vertical) or top/bottom (horizontal).

I've also modified "viz_ui_slider.py" with tutorials respective to horizontal and left orientation and have updated tests and tutorials

### CODE REVIEW

File: fury/optpkg.py [hunk 1]
```diff
@@ -44,7 +44,8 @@ class TripWire(object):
     ...     import silly_module_name
     ... except ImportError:
     ...    silly_module_name = TripWire('We do not have silly_module_name')
-    >>> silly_module_name.do_silly_thing('with silly string') #doctest: +IGNORE_EXCEPTION_DETAIL
+    >>> silly_module_name.do_silly_thing('with silly string') #doctest:
+                                +IGNORE_EXCEPTION_DETAIL
```
Comment (skoudoro) [inline fury/optpkg.py]: Unfortunately, you can not change this line. it will make the test fail.

File: fury/ui.py [hunk 2]
```diff
@@ -1793,7 +1793,8 @@ def __init__(self, center=(0, 0),
                  inner_radius=0, outer_radius=10, handle_side=20,
                  font_size=16,
                  orientation="horizontal",
-                 text_template="{value:.1f} ({ratio:.0%})", shape="disk"):
+                 text_template="{value:.1f} ({ratio:.0%})", shape="disk",
+                 alingment=''):
```
Comment (skoudoro) [inline fury/ui.py]: - typo: `alignment` instead

File: fury/ui.py [hunk 3]
```diff
@@ -1832,14 +1833,22 @@ def __init__(self, center=(0, 0),
         self.orientation = orientation.lower()
         self.default_color = (1, 1, 1)
         self.active_color = (0, 0, 1)
+        self.alingment = alingment.lower()
```
Comment (skoudoro) [inline fury/ui.py]: typo: `alignment`, please fix it everywhere

File: fury/ui.py [hunk 4]
```diff
@@ -1832,14 +1833,22 @@ def __init__(self, center=(0, 0),
...
         super(LineSlider2D, self).__init__()
-
         if self.orientation == "horizontal":
             self.track.width = length
             self.track.height = line_width
+            self.alingment = 'bottom' if not self.alingment else self.alingment
+            if self.alingment not in ['top', 'bottom']:
+                raise ValueError("""Unknown alingment: choose from 'top'
+                                or 'bottom'""")
```
Comment (skoudoro) [inline fury/ui.py]: indentation problem

File: fury/ui.py [hunk 5]
```diff
@@ -1861,7 +1870,7 @@ def __init__(self, center=(0, 0),
 
         self.value = initial_value
         self.update()
-
+
```
Comment (skoudoro) [inline fury/ui.py]: space to remove

File: fury/optpkg.py (hunk 1 above)
Comment (PaulNicolasHunter) [reply fury/optpkg.py]: Great, I'll add more test

File: fury/ui.py (hunk 2 above)
Comment (PaulNicolasHunter) [reply fury/ui.py]: noted

File: fury/optpkg.py (hunk 1 above)
Comment (PaulNicolasHunter) [reply fury/optpkg.py]: @skoudoro I've changed the line back to default, but still, the test sare failing, what do you think could be the issue?

**Comment by PaulNicolasHunter** [comment]
hi, @skoudoro I've made the necessary changes as you suggested.
kindly have a look.

regards,

**Comment by skoudoro** [comment]
Hi @PaulNicolasHunter,

Thank you first for this first update. Can you please address all my comments? As a reminder, I am ust citing myself below: 

>  Can you fix all pep8 issue, and add some unittests. You can look at the slider unittest in test_ui.py.

**Comment by PaulNicolasHunter** [comment]
> Hi @PaulNicolasHunter,
> 
> Thank you first for this first update. Can you please address all my comments? As a reminder, I will just cite myself below:
> 
> > Can you fix all pep8 issue, and add some unittests. 

I think the new report shows that there aren't any pep8 issues.

> > You can look at the slider unittest in test_ui.py.
Sure.

**Comment by skoudoro** [comment]
> I think the new report shows that there aren't any pep8 issues.

there are 2 reports. The second one is the second message from this thread. here the link: https://github.com/fury-gl/fury/pull/189#issuecomment-596170376

File: docs/tutorials/01_ui/viz_ui_slider.py [hunk 6]
```diff
@@ -97,10 +137,10 @@ def translate_cube(slider):
 show_manager.scene.set_camera(position=(0, 0, 150))
 show_manager.scene.reset_clipping_range()
 show_manager.scene.azimuth(30)
-interactive = False
+interactive = True
```
Comment (skoudoro) [inline docs/tutorials/01_ui/viz_ui_slider.py]: Can you put it back to `False` ? thank you

File: docs/tutorials/01_ui/viz_ui_slider.py [hunk 7]
```diff
@@ -97,10 +137,10 @@ def translate_cube(slider):
...
 
 if interactive:
     show_manager.start()
 
-window.record(show_manager.scene,
-              size=current_size, out_path="viz_slider.png")
+# window.record(show_manager.scene,
+#               size=current_size, out_path="viz_slider.png")
```
Comment (skoudoro) [inline docs/tutorials/01_ui/viz_ui_slider.py]: Can you put it back and uncomment? Thank you

File: fury/tests/test_ui.py [hunk 8]
```diff
@@ -701,7 +767,6 @@ def test_ui_listbox_2d(interactive=False):
 
     # Values that will be displayed by the listbox.
     values = list(range(1, 42 + 1))
-    values.append("A Very Very Long Item To Test Text Overflow of List Box 2D")
```
Comment (skoudoro) [inline fury/tests/test_ui.py]: you should keep this

File: fury/tests/test_ui.py [hunk 9]
```diff
@@ -723,11 +788,11 @@ def test_ui_listbox_2d(interactive=False):
     #  2. Ctrl + click on 2,
     #  3. Ctrl + click on 2.
     #  4. Use scroll bar to scroll to the bottom.
-    #  5. Click on "A Very Very Long Item...".
```
Comment (skoudoro) [inline fury/tests/test_ui.py]: you should keep this

File: fury/tests/test_ui.py [hunk 10]
```diff
@@ -723,11 +788,11 @@ def test_ui_listbox_2d(interactive=False):
...
+    #  5. Click on 42.
     #  6. Use scroll bar to scroll to the top.
     #  7. Click on 1
     #  8. Use mouse wheel to scroll down.
-    #  9. Shift + click on "A Very Very Long Item...".
```
Comment (skoudoro) [inline fury/tests/test_ui.py]: you should keep this

File: fury/tests/test_ui.py [hunk 11]
```diff
@@ -757,8 +822,7 @@ def _on_change():
     event_counter.check_counts(expected)
 
     # Check if the right values were selected.
-    expected = [[1], [1, 2], [1], ["A Very Very Long Item To \
-Test Text Overflow of List Box 2D"], [1], values]
```
Comment (skoudoro) [inline fury/tests/test_ui.py]: you should keep this

File: fury/tests/test_ui.py [hunk 12]
```diff
@@ -768,9 +832,7 @@ def _on_change():
     show_manager.play_events_from_file(recording_filename)
 
     # Check if the right values were selected.
-    expected = [[1], [2], [2], ["A Very Very Long Item To \
-Test Text Overflow of List Box 2D"], [1], ["A Very Very Long Item To Test \
-Text Overflow of List Box 2D"]]
```
Comment (skoudoro) [inline fury/tests/test_ui.py]: you should keep this

File: fury/ui.py [hunk 13]
```diff
@@ -1829,20 +1830,29 @@ def __init__(self, center=(0, 0),
...
         elif self.orientation == "vertical":
+            self.alignment = 'left' if not self.alignment else self.alignment
             self.track.width = line_width
             self.track.height = length
         else:
             raise ValueError("Unknown orientation")
 
+        if self.alignment not in self.align_dict[self.orientation]:
+            raise ValueError("Unknown alignment: choose from '{}' or '{}'".
+                             format(*self.align_dict[self.orientation]))
```
Comment (skoudoro) [inline fury/ui.py]: Nice!

File: fury/ui.py [hunk 14]
```diff
@@ -3447,14 +3460,14 @@ def _setup(self):
             self.scroll_bar, size - self.scroll_bar.size - self.margin)
 
         # Initialisation of empty text actors
-        self.slot_width = size[0] - self.scroll_bar.size[0] - \
```
Comment (skoudoro) [inline fury/ui.py]: you should keep this

File: fury/ui.py [hunk 15]
```diff
@@ -3447,14 +3460,14 @@ def _setup(self):
...
+        slot_width = size[0] - self.scroll_bar.size[0] - \
             2 * self.margin - self.margin
         x = self.margin
         y = size[1] - self.margin
         for _ in range(self.nb_slots):
             y -= self.slot_height
             item = ListBoxItem2D(list_box=self,
-                                 size=(self.slot_width, self.slot_height),
```
Comment (skoudoro) [inline fury/ui.py]: you should keep this

File: fury/ui.py [hunk 16]
```diff
@@ -3652,16 +3665,7 @@ def update(self):
...
-            char_width = slot.textblock.size[0] - self.margin
-            permissible_chars = int(self.slot_width)//char_width
-            total_chars = len(str(choice))
-            if total_chars > permissible_chars:
-                excess_chars = total_chars - permissible_chars
-                wrapped_choice = choice[:(-excess_chars) + 3] + "..."
-                slot.element = choice
-                slot.textblock.message = wrapped_choice
-            else:
-                slot.element = choice
```
Comment (skoudoro) [inline fury/ui.py]: you should keep this whole block

File: docs/tutorials/01_ui/viz_ui_slider.py (hunk 6 above)
Comment (PaulNicolasHunter) [reply docs/tutorials/01_ui/viz_ui_slider.py]: got it.

File: fury/ui.py (hunk 16 above)
Comment (PaulNicolasHunter) [reply fury/ui.py]: Wow, I really don't remember removing any lines from the test_ui.py or ui.py.
but thanks for pointing out I'll look into it.

**Comment by PaulNicolasHunter** [comment]
Hi @skoudoro I not able to understand why the tests are failing at fury-gl.fury and Travis, I tried digging at details but couldn't make any sense of it, this is my first time using this, can you please give me any hints, please?

thanks!

**Comment by skoudoro** [comment]
Hi @PaulNicolasHunter,

The errors are not your fault so no worries, I will restart those CI's.

However the pep8. look at the following link to see them: https://github.com/fury-gl/fury/pull/189#issuecomment-596170376

**Comment by PaulNicolasHunter** [comment]
Hey, @skoudoro I've fixed those pep8 issues.
let me know if there is any other change needed.

File: docs/tutorials/01_ui/viz_ui_slider.py [hunk 17]
```diff
@@ -37,13 +37,43 @@ def cube_maker(color=(1, 1, 1), size=(0.2, 0.2, 0.2), center=(0, 0, 0)):
 cube = cube_maker(color=(0, 0, 1), size=(20, 20, 20), center=(15, 0, 0))
 
 ###############################################################################
-# Now we'll add two sliders: one circular and one linear.
+# Now we'll add two sliders: 1 circular and 4 linear sliders.
```
Comment (skoudoro) [inline docs/tutorials/01_ui/viz_ui_slider.py]: `two sliders` -> `five sliders`

File: docs/tutorials/01_ui/viz_ui_slider.py [hunk 18]
```diff
@@ -37,13 +37,43 @@ def cube_maker(color=(1, 1, 1), size=(0.2, 0.2, 0.2), center=(0, 0, 0)):
...
+                                             orientation='vertical',
+                                             min_value=-10, max_value=10,
+                                             alignment='right')
+
+###############################################################################
+# By default the alignment are 'bottom' for horizontal and 'top' for vertical
+
+hor_line_slider = ui.LineSlider2D(center=(650, 100), initial_value=0,
+                                  orientation='horizontal', min_value=-10,
+                                  max_value=10)
```
Comment (skoudoro) [inline docs/tutorials/01_ui/viz_ui_slider.py]: slider to remove, not needed

File: docs/tutorials/01_ui/viz_ui_slider.py [hunk 19]
```diff
@@ -37,13 +37,43 @@ def cube_maker(color=(1, 1, 1), size=(0.2, 0.2, 0.2), center=(0, 0, 0)):
...
+
+ver_line_slider = ui.LineSlider2D(center=(700, 250), initial_value=0,
+                                  orientation='vertical', min_value=-10,
+                                  max_value=10)
```
Comment (skoudoro) [inline docs/tutorials/01_ui/viz_ui_slider.py]: slider to remove, not needed

File: docs/tutorials/01_ui/viz_ui_slider.py [hunk 20]
```diff
@@ -37,13 +37,43 @@ def cube_maker(color=(1, 1, 1), size=(0.2, 0.2, 0.2), center=(0, 0, 0)):
...
+                                            alignment='left')
+
+ver_line_slider_text_right = ui.LineSlider2D(center=(150, 400),
+                                             initial_value=0,
+                                             orientation='vertical',
+                                             min_value=-10, max_value=10,
+                                             alignment='right')
+
+###############################################################################
+# By default the alignment are 'bottom' for horizontal and 'top' for vertical
```
Comment (skoudoro) [inline docs/tutorials/01_ui/viz_ui_slider.py]: You can move this text to the section above. after the following text: `4 linear sliders.`

File: docs/tutorials/01_ui/viz_ui_slider.py [hunk 21]
```diff
@@ -67,7 +97,12 @@ def translate_cube(slider):
     cube.SetPosition(value, 0, 0)
 
 
-line_slider.on_change = translate_cube
+hor_line_slider_text_top.on_change = translate_cube
+hor_line_slider_text_bottom.on_change = translate_cube
+ver_line_slider_text_left.on_change = translate_cube
+ver_line_slider_text_right.on_change = translate_cube
+hor_line_slider.on_change = translate_cube
+ver_line_slider.on_change = translate_cube
```
Comment (skoudoro) [inline docs/tutorials/01_ui/viz_ui_slider.py]: same, you can remove `hor_line_slider` and `ver_line_slider`

File: docs/tutorials/01_ui/viz_ui_slider.py [hunk 22]
```diff
@@ -81,14 +116,25 @@ def translate_cube(slider):
 
 show_manager.scene.add(cube)
 show_manager.scene.add(ring_slider)
-show_manager.scene.add(line_slider)
+show_manager.scene.add(hor_line_slider_text_top)
+show_manager.scene.add(hor_line_slider_text_bottom)
+show_manager.scene.add(ver_line_slider_text_left)
+show_manager.scene.add(ver_line_slider_text_right)
+show_manager.scene.add(hor_line_slider)
+show_manager.scene.add(ver_line_slider)
```
Comment (skoudoro) [inline docs/tutorials/01_ui/viz_ui_slider.py]: same, you can remove `hor_line_slider` and `ver_line_slider`

File: docs/tutorials/01_ui/viz_ui_slider.py [hunk 23]
```diff
@@ -81,14 +116,25 @@ def translate_cube(slider):
...
 
 cube.SetVisibility(True)
 ring_slider.set_visibility(True)
-line_slider.set_visibility(True)
+hor_line_slider_text_top.set_visibility(True)
+hor_line_slider_text_bottom.set_visibility(True)
+ver_line_slider_text_left.set_visibility(True)
+ver_line_slider_text_right.set_visibility(True)
+hor_line_slider.set_visibility(True)
+ver_line_slider.set_visibility(True)
```
Comment (skoudoro) [inline docs/tutorials/01_ui/viz_ui_slider.py]: same, you can remove `hor_line_slider` and `ver_line_slider`

Comment (PaulNicolasHunter) [reply docs/tutorials/01_ui/viz_ui_slider.py]: > Ok, the last review, I just tested it and it works well. After your last update, I will merge your PR.
> 
> Any other comments on this PR?

Got it, please do let me know if any further changes needed.
cheers.

**Comment by skoudoro** [comment]
Thank you @PaulNicolasHunter for this work! merging

**Pull request merged**